### PR TITLE
[YUNIKORN-1163] Return history of application state transitions in REST API

### PR DESCRIPTION
**Pull request by craigcondit**
### What is this PR for?
Expose application state transitions in the REST API as part of the application objects.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1163

### How should this be tested?
Unit tests updated to verify event logs.

### Screenshots (if appropriate)
stateLog

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Review by yangwwei — APPROVED** [review]
LGTM, this will be very useful, thank you @craigcondit !

**Comment by craigcondit** [comment]
@wilfred-s I know you had some ideas surrounding this as well. Is this what you had in mind?

File: pkg/scheduler/objects/application.go [hunk 1]
```diff
@@ -53,6 +53,11 @@ const (
 	AppTagStateAwareDisable string = "application.stateaware.disable"
 )
 
+type EventLogEntry struct {
+	EventTime        time.Time
+	ApplicationState string
+}
```
Comment (wilfred-s) [inline pkg/scheduler/objects/application.go:59]: This should be called StateLogEntry, we log state changes not every event.

File: pkg/scheduler/objects/application.go [hunk 2]
```diff
@@ -80,6 +85,7 @@ type Application struct {
 	gangSchedulingStyle  string                 // gang scheduling style can be hard (after timeout we fail the application), or soft (after timeeout we schedule it as a normal application)
 	finishedTime         time.Time              // the time of finishing this application. the default value is zero time
 	rejectedMessage      string                 // If the application is rejected, save the rejected message
+	eventLog             []*EventLogEntry       // event log for this application
```
Comment (wilfred-s) [inline pkg/scheduler/objects/application.go:88]: This should be called stateLog

File: pkg/scheduler/objects/application.go [hunk 3]
```diff
@@ -142,6 +149,20 @@ func (sa *Application) SetState(state string) {
 	sa.stateMachine.SetState(state)
 }
 
+func (sa *Application) recordEvent(time time.Time, appState string, prevState string) {
```
Comment (wilfred-s) [inline pkg/scheduler/objects/application.go:152]: signature cleanup:
* time is always passed in as time.Now(), we can set it here. The nano second differences are not important.
* prevState is not used as the ordering provides the previous state already

File: pkg/scheduler/objects/application.go [hunk 4]
```diff
@@ -281,6 +303,9 @@ func (sa *Application) timeoutStateTimer(expectedState string, event application
 				sa.notifyRMAllocationReleased(sa.rmID, toRelease, si.TerminationType_TIMEOUT, "releasing placeholders on app complete")
 				sa.clearStateTimer()
 			} else {
+				// HandleApplicationEvent assumes the lock to be held
+				sa.Lock()
+				defer sa.Unlock()
```
Comment (wilfred-s) [inline pkg/scheduler/objects/application.go:308]: We have two choices: 
1. move adding this lock and a lock review of `timeoutStateTimer()` to a new jira., or
2. fix all locking issues here: `getPlaceholderAllocations()` accesses the allocations map unlocked which might cause issues. `clearStateTimer()` in all cases but here is called under lock.
We discussed the `clearStateTimer()` call yesterday and I am not sure it causes a problem but we might as well be careful.

**Review by wilfred-s — CHANGES_REQUESTED** [review]

File: pkg/scheduler/objects/application.go (hunk 1 above)
Comment (craigcondit) [reply pkg/scheduler/objects/application.go]: Updated in new commit.

File: pkg/scheduler/objects/application.go (hunk 2 above)
Comment (craigcondit) [reply pkg/scheduler/objects/application.go]: Updated in new commit.

**Review by pbacsko — APPROVED** [review]
+1 LGTM

**Pull request closed without merging**